### PR TITLE
[pickers] Stop root exporting `locales`

### LINKED ISSUE #10918: [l10n] Stop re-exporting `locales` from package root
We currently re-export locales from in the root `index` file so that they are available from both `import { frFR } from '@mui/x-date-pickers/locales'` and `import { frFR } from '@mui/x-date-pickers'`.
https://github.com/mui/mui-x/blob/947619da1c74dcffc15f79ae79c00a92a21f0fd1/packages/x-date-pickers/src/index.ts#L8

We should avoid root-level exporting as it doesn't solve any particular problem besides making the package seem bulkier.

Related to https://github.com/mui/material-ui/issues/39525

```diff
-import { frFR } from '@mui/x-date-pickers';
+import { frFR } from '@mui/x-date-pickers/locales';
```

**Search keywords**:

### PR DESCRIPTION
**Pull request by LukasTy**
Fixes #10918

- Remove root `locales` re-exporting
  - `import { frFR } from '@mui/x-date-pickers';` would no longer work after this, but we never endorse such usage in the docs anyway.
- Keep root re-exporting `l10n` related types.
- Re-export `locales` via `@mui/x-date-pickers-pro/locales`.

Let's review the approach, if we are fine with it, then I'll:
- [x] Add migration guide entry
- [ ] ~Add codemod~ Don't see much point in having it given all circumstances

### CODE REVIEW

**Review by michelengelen — APPROVED** [review]
LGTM

**Review by flaviendelangle — APPROVED** [review]
OK to go with it since the discussion for depth 3 are far from over :ok_hand:

**Comment by flaviendelangle** [comment]
We just need some migration guide and maybe codemod before merging

**Comment by flaviendelangle** [comment]
> I have given some thought about adding a codemod and concluded that IMHO it's not worth the effort to maybe replace 1 import for 1 in 100 user cases.
If we had encouraged using root-level imports previously, I think it could make sense to add it, but given our current documentation, I don't think it's worth it. 🤷 🙈
WDYT? 

I'd be in favor of doing the codemod because it's fairly easy to do and adding a green check on the migration guide improves the overall feeling that the migration is easy.
But if you think it's wasted time I won't fight to have it :laughing:

**Comment by LukasTy** [comment]
> I'd be in favor of doing the codemod because it's fairly easy to do and adding a green check on the migration guide improves the overall feeling that the migration is easy.

I would agree with you and kind of think that in the Data Grid case, it might make sense (however, they didn't add one, at least for now) because their demos had examples of importing locales from the root. 🤔 
If I understand correctly. any new codemod means yet another iteration over every file in a given project. Because of that, in my head, it just means wasted resources and time on someone else's end. 🙈 🤷

File: docs/data/migration/migration-pickers-v6/migration-pickers-v6.md [hunk 1]
```diff
@@ -429,6 +429,19 @@ dayjs.extend(customParseFormatPlugin);
...
+
+The `locales` export has been removed from the root of the packages.
+In an effort to reduce the bundle size, the locales are now only available from the `@mui/x-date-pickers/locales` or `@mui/x-date-pickers-pro/locales` paths.
+If you were still relying on the root level export, please update your code.
+
+Before v7, it was possible to import locales from the package root (i.e. `import { frFR } from '@mui/x-date-pickers'`).
+
+```diff
+- import { frFR } from '@mui/x-date-pickers';
++ import { frFR } from '@mui/x-date-pickers/locales';
```
Comment (oliviertassinari) [inline docs/data/migration/migration-pickers-v6/migration-pickers-v6.md:442]: Same as https://github.com/mui/mui-x/pull/11650#discussion_r1448084598

```suggestion
-import { frFR } from '@mui/x-date-pickers';
+import { frFR } from '@mui/x-date-pickers/locales';
```

Comment (LukasTy) [reply docs/data/migration/migration-pickers-v6/migration-pickers-v6.md:442]: Thanks for bringing it up.
Raised a PR fixing it and some other tiny issues. 😉 
https://github.com/mui/mui-x/pull/11695

**Comment by oliviertassinari** [comment]
Bundle size in https://bundlephobia.com/ is now smaller, cool



https://bundlephobia.com/package/@mui/x-date-pickers@7.0.0-alpha.8



https://bundlephobia.com/package/@mui/x-date-pickers@7.0.0-alpha.9

**Comment by LukasTy** [comment]
> Bundle size in https://bundlephobia.com/ is now smaller, cool

Nice!. 👏 
The difference on Data Grid side is even greater, because their localisation files are bigger.
Screenshot 2024-01-25 at 11 04 04

**Comment by oliviertassinari** [comment]
Now, to be fair, people who use https://marketplace.visualstudio.com/items?itemName=wix.vscode-import-cost, will see an increase in v7 vs. v6 


https://bundlephobia.com/package/@mui/x-date-pickers@6.19.0
Screenshot 2024-01-25 at 14 12 12

https://bundlephobia.com/package/@mui/x-date-pickers@7.0.0-alpha.9
Screenshot 2024-01-25 at 14 11 53

**Comment by flaviendelangle** [comment]
For the pickers, I think it's due to our double textfield approach, we bundle both the TreeView and the new component.

**Comment by oliviertassinari** [comment]
Connecting this to #5550

**Pull request merged**